tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -241,12 +241,34 @@ impl ServerKey {
         self.unchecked_add_assign_parallelized_work_efficient(lhs, rhs);
     }
 
-    pub(crate) fn is_eligible_for_parallel_carryless_add(&self) -> bool {
+    pub(crate) fn is_eligible_for_parallel_single_carry_propagation<T>(&self, ct: &T) -> bool
+    where
+        T: IntegerRadixCiphertext,
+    {
         // having 4-bits is a hard requirement
         // as the parallel implementation uses a bivariate BPS where individual values need
         // 2 bits
         let total_modulus = self.key.message_modulus.0 * self.key.carry_modulus.0;
-        total_modulus >= (1 << 4)
+        let has_enough_bits_per_block = total_modulus >= (1 << 4);
+        if !has_enough_bits_per_block {
+            return false;
+        }
+
+        // The fully parallelized way introduces more work
+        // and so is slower for low number of blocks
+        const MIN_NUM_BLOCKS: usize = 6;
+        let has_enough_blocks = ct.blocks().len() >= MIN_NUM_BLOCKS;
+        if !has_enough_blocks {
+            return false;
+        }
+
+        // Use rayon to get that number as the implementation uses rayon for parallelism
+        let has_enough_threads = rayon::current_num_threads() >= ct.blocks().len();
+        if !has_enough_threads {
+            return false;
+        }
```

Comment:
do we use a single thread per block in the implementation ?

Reply:
The parallel carry propagation is done in log2(num_block) steps, where the first step start wth num_block threads
then each following iteration divides the requires number of blocks by 2.

On my laptop, this checks makes me use the non parallel version of the algorithms for all bit sizes because I do not have enough threads, and this makes the operations (add/sub/neg) between 25% and 45% faster

Reply:
Ok ! I did not recall how many threads were required for each step of the algo, thanks